### ISSUE TITLE
Supervised byte-level family classifier with SHAP explanations (the thing ADR-019 says the RE stage cannot do)

### ISSUE THREAD

**Issue by chrisshaiman** [chrisshaiman/lamware#376]
A supervised byte-level classifier that produces a family label with a **per-prediction explanation**, as a fast stage running before Ghidra.

This exists because of ADR-019: the RE stage **cannot** do family attribution — an LLM reading decompiled code of a packed binary, over an open family set, has no signal to work with (qwen 0/14, the Claude reference 0/7). But that is a statement about *that method*, not about the task. Supervised classifiers over byte-level features reach ~91% on packed malware, because a packer stub is generic **as source code** while remaining statistically distinctive **as bytes**.

So: get the family from the bytes, and let the RE stage do what it is actually good at.

## What it is

Never looks at code. Extracts a fixed-length numeric vector from the raw PE and classifies it:

- **Byte histogram** — frequency of each of the 256 byte values
- **Byte-entropy histogram** — entropy over a sliding window, 2D-binned against byte value. This is what makes packing *legible* rather than opaque: packers leave distinct entropy fingerprints
- **PE structural features** — section count/names/sizes/entropy, import table shape, exports, resources, timestamps, debug info
- **String statistics** — count, mean length, printable distribution, path/URL/registry-like counts

Model is gradient-boosted trees (LightGBM). Training is minutes on CPU; inference is milliseconds, negligible against a 40-minute Ghidra run.

## What to import (checked 2026-08-10)

| source | licence | notes |
|---|---|---|
| **[FutureComputing4AI/EMBER2024](https://github.com/FutureComputing4AI/EMBER2024)** | **Apache-2.0** | **Preferred.** 6,787 families (2,538 with ≥10 instances). Ships 14 LightGBM benchmark classifiers across detection, **family ID**, and tag prediction, plus feature-extraction and training code. Same group as MOTIF. |
| [boozallen/MOTIF](https://github.com/boozallen/MOTIF) `benchmarks/lgbm.py` | Booz Allen PL v1.0 | Ready-made LightGBM over MOTIF's 454 families. Internal use permitted; **do not commit samples or derived artefacts to this public repo**. |
| ~~elastic/ember~~ | "Other" / NOASSERTION | **Avoid.** Archived Nov 2024, non-standard licence — awkward for a public repo. The feature *format* is still the de-facto standard. |

EMBER2024 removes the limitation that made this unattractive earlier: 6,787 current families rather than 454 frozen at 2020, so latrodectus / rhadamanthys / warmcookie are potentially in scope rather than structurally unlearnable.

## It explains itself — exactly, not approximately

**SHAP TreeExplainer is exact for gradient-boosted trees.** Per-prediction contributions sum to the output, and EMBER's features are named and largely analyst-legible:

```
amadey  0.61
  +0.31  section .rsrc entropy = 7.91         (encrypted resource)
  +0.18  imports from WS2_32.dll = 12         (network capable)
  +0.12  byte-entropy bin [6.5-7.0] elevated  (packed region)
  -0.07  has_debug = 0
```

SHAP + LightGBM on EMBER features is established practice, not novel work.

**Attribution is not reasoning.** It says what the model keyed on, not why that means Amadey. Better than a bare label; still an assertion rather than an argument.

## The caveat that makes explainability load-bearing

["Beyond the Wrapper: Identifying Artifact Reliance in Static Malware Classifiers using TRUSTEE"](https://arxiv.org/pdf/2605.07034) documents these classifiers keying on **packer/wrapper artefacts rather than family traits**.

That is the same mechanism that makes them work on packed samples — and their fragility, because crypters are shared across families and rotate. A model that is really a packer detector wearing a family label looks excellent until the crypter changes.

**So the SHAP output is not a nicety, it is the detector for that failure mode.** If the top features for every Amadey prediction are entropy bins with zero import-table contribution, we have built a packer classifier and should say so.

Acceptance should include: inspect SHAP feature distributions per family and assert that predictions are not driven solely by entropy/packing features.

## The hybrid worth building instead of a bare label

Rather than emitting a label beside the analysis, feed the top SHAP features into the RE stage as **evidence**:

> Classifier: amadey (0.61), driven by `.rsrc` entropy 7.91, 12 WS2_32 imports, elevated entropy in the 6.5-7.0 band.

The model can then corroborate or **dispute** it against the actual code — and a dispute is a *grounded claim*, which is already measurable. That turns an unexplainable label into a hypothesis the RE stage can test, fitting the grounding philosophy rather than sitting beside it.

**Caution:** this reintroduces the ADR-019 label-leak concern deliberately. If the classifier's guess is in the prompt, `family_guess` stops being a contamination probe and becomes a measure of whether the model agrees with the classifier. Any eval arm using this hybrid must be scored separately, and the plain arm kept for the probe.

## Sketch

```
sample ──> EMBER feature extraction ──> LightGBM ──> {family, confidence, shap_top_n}
                                                              │
                                    report["family_classifier"] = {...}
                                                              │
                              (optional) ──> RE stage as a stated hypothesis
```

- New `family-classifier` role: container with LIEF/pefile + LightGBM, model baked in
- Model trained offline once; artefact is a few MB, pinned by sha256 like every other build input (#344)
- Runs in parallel with triage, before Ghidra
- Emits a **probability distribution**, not an argmax

## Limitations to state in the output, not discover later

- **Closed set.** It can only name families it was trained on. An unseen family is confidently assigned to the nearest known one — so an explicit "unknown" confidence threshold is required, and the argmax must never be taken blindly.
- **Concept drift.** Packers change; a model trained on today's corpus degrades. Needs periodic retraining and a drift signal.
- **No reasoning.** Suitable for triage routing and as a stated hypothesis; not as an analyst-facing conclusion on its own.

## Human effort

**Not per-sample review.** Up-front only:

1. Calibrate the confidence threshold on a held-out split (once)
2. Retrain when drift appears (occasional)
3. Optionally work a disagreement queue — classifier says X, MalwareBazaar says Y, RE stage says unknown

## Sequencing

**After unpacking.** Running Ghidra over CAPE's dropped/unpacked payloads (`/opt/CAPEv2/storage/analyses/<task>/dropped`, already read by the investigate tools) improves grounding, capability claims, *and* family attribution by the RE stage itself, because config markers and distinctive strings survive in a dump. That changes the numbers this project has been measuring for weeks; this feature adds a new one.

Strong second feature, not the first.

## Related

ADR-019 (`docs/DECISIONS.md`) — why the RE stage cannot do this. #314 — ground truth. #368 — MOTIF corpus.
